### DISCUSSION TITLE
Why fine tuning a very good model worsened it ?

### DISCUSSION THREAD

**Discussion by Ca-ressemble-a-du-fake** [coqui-ai/TTS#2021] [General Q&A] [answered]
Hi,

I have trained a first model in French (single speaker Robert, VITS) which is very good at 60 steps (dataset with 5k samples).

I have also trained another model in French (single speaker Frank, VITS as well) which is not as good as the previous one but good enough to be intelligible at 90 k steps (dataset with 520 samples).

Consequently to improve the quality of the second model I decided to fine tune the first (very good) one with the dataset from the second one. Yet it has worsened the results after 10 k steps and even after 20 k steps of fine tuning steps. As @WeberJulian  advised several times I froze DP and encoder (other model parameters were left untouched). I decreased the learning rate from 1E-3 to 1E-5 not to loose the previous learning. The 

Here is an example of the fine tuned model for "Il m'a fallu du temps pour créer cette voix alors je ne vais pas me taire" (recorded from my phone since I could not export the test audio from Tensorboard🤪, so excuse for the sound quality) :
https://user-images.githubusercontent.com/91517923/192355904-49f6f1f3-7fc8-4184-8dae-a5c749ed37a5.mp4
The voice sounds distorted.

Should I let it continue up to eg 60 k more steps ? => after 40 k steps model output did NOT improve.

Or should I restart the fine tuning without freezing encoder and DP ?

How can one knows when to freeze DP and encoder or something else while finetuning ?

Thank you in advance for your advices!

**Comment by WeberJulian** [comment, marked as answer]
You trained a VITS model from scratch with only 520 samples and it produced intelligible speech for test sentences?

**Reply by Ca-ressemble-a-du-fake** [reply]
Yes indeed. With exactly 563 samples ranging from 1 to 10 seconds. During inference the speech was fully intelligible, the speaker could be recognized (with effort), but the voice sounded a bit robotic. 
I try to unfreeze DP (freeze_DP = false) while keeping freenze_encoder = true with the same poor results. 
I tried to unfreeze encoder and DP and it looked promising in the beginning but now after 20 k steps it sounds poor again. The speech is correct (intelligible and close to speaker) only for the first words :  "Il m'a fallu" is correct and then it is only garbage (it sounds like a drunk man for "du temps pour créer cette voix alors je ne vais pas me taire").

**Reply by WeberJulian** [reply]
Yeah I only tried freezing in multi-speaker regimen, maybe it doesn't work when the text encoder is not general enough.
Maybe you can try fine tuning from the VCTK checkpoint with both french datasets datasets (and maybe mailabs?).

**Reply by Ca-ressemble-a-du-fake** [reply]
I attach a new sample at step 100 k (60 k of original model + 40 k of fine tuning) maybe the sound will give you a clue of what's happening! The corresponding text sentence is "il m'a fallu du temps pour créer cette voix alors je ne vais pas me taire.".

https://user-images.githubusercontent.com/91517923/192682261-e3adfd85-716f-45c2-8d18-0e42379466d8.mp4

So it starts with "il m'a fallu" that can be understood.

But now if I truncate it to only the beginning "Il m'a fallu du temps."
It now starts saying "Je va fallu".

https://user-images.githubusercontent.com/91517923/192682350-6673d3e7-0461-4e4a-b331-b81b40ec82c8.mp4

Maybe the characters in metada.csv are not encoded the same way. In the original model I remember normalizing them because it came from Word under Windows, and I did not do it for the small dataset.

**Reply by WeberJulian** [reply]
I don't really know, could you upload the config please ?

**Reply by Ca-ressemble-a-du-fake** [reply]
Sure here you are Sir !
```
  {
    "output_path": "/home/caraduf/VoiceCloning/Models/FineTuning/A_to_B",
    "logger_uri": null,
    "run_name": "VITS_FINE_TUNING_A_TO_B",
    "project_name": null,
    "run_description": "Fine tuning very good model A to speaker B",
    "print_step": 25,
    "plot_step": 100,
    "model_param_stats": false,
    "wandb_entity": null,
    "dashboard_logger": "tensorboard",
    "log_model_step": null,
    "save_step": 10000,
    "save_n_checkpoints": 5,
    "save_checkpoints": true,
    "save_all_best": false,
    "save_best_after": 10000,
    "target_loss": null,
    "print_eval": false,
    "test_delay_epochs": 0,
    "run_eval": true,
    "run_eval_steps": null,
    "distributed_backend": "nccl",
    "distributed_url": "tcp://localhost:54321",
    "mixed_precision": true,
    "epochs": 30001,
    "batch_size": 64,
    "eval_batch_size": 16,
    "grad_clip": [
        5.0,
        5.0
    ],
    "scheduler_after_epoch": true,
    "lr": 0.00001,
    "optimizer": "AdamW",
    "optimizer_params": {
        "betas": [
            0.8,
            0.99
        ],
        "eps": 1e-09,
        "weight_decay": 0.01
    },
    "lr_scheduler": "",
    "lr_scheduler_params": {},
    "use_grad_scaler": false,
    "cudnn_enable": true,
    "cudnn_deterministic": false,
    "cudnn_benchmark": false,
    "training_seed": 54321,
    "model": "vits",
    "num_loader_workers": 4,
    "num_eval_loader_workers": 0,
    "use_noise_augment": false,
    "audio": {
        "fft_size": 1024,
        "sample_rate": 16000,
        "win_length": 1024,
        "hop_length": 256,
        "num_mels": 80,
        "mel_fmin": 0,
        "mel_fmax": null
    },
    "use_phonemes": true,
    "phonemizer": "espeak",
    "phoneme_language": "fr-fr",
    "compute_input_seq_cache": false,
    "text_cleaner": "phoneme_cleaners",
    "enable_eos_bos_chars": false,
    "test_sentences_file": "",
    "phoneme_cache_path": "Phonemes_cache",
    "characters": {
        "characters_class": "TTS.tts.models.vits.VitsCharacters",
        "vocab_dict": null,
        "pad": "_",
        "eos": "",
        "bos": "",
        "blank": null,
        "characters": "ABCDEFGHIJLMNOPQRSTUVWXYabcdefghijklmnopqrstuvwxyz\u0153",
        "punctuations": " !(),-.:;?\u00ab\u00bb\u2013\u2019",
        "phonemes": " !,-.:;?abdefhijklmnopstuvwyz\u00ab\u00bb\u00f0\u00f8\u014b\u0153\u0251\u0252\u0254\u0259\u025b\u025c\u0261\u026a\u0272\u0279\u0281\u0283\u028a\u028c\u0292\u02c8\u02cc\u02d0\u0303\u03b8",
        "is_unique": true,
        "is_sorted": true
    },
    "add_blank": true,
    "batch_group_size": 0,
    "loss_masking": null,
    "min_audio_len": 1,
    "max_audio_len": Infinity,
    "min_text_len": 1,
    "max_text_len": Infinity,
    "compute_f0": false,
    "compute_linear_spec": true,
    "precompute_num_workers": 0,
    "start_by_longest": true,
    "datasets": [
        {
            "name": "caraduf",
            "path": "/home/caraduf/VoiceCloning/Speaker_B_dataset",
            "meta_file_train": "metadata.csv",
            "ignored_speakers": null,
            "language": "fr-fr",
            "meta_file_val": "",
            "meta_file_attn_mask": ""
        }
    ],
    "test_sentences": [
        [
            "Il m'a fallu du temps pour cr\u00e9er cette voix alors je ne vais pas me taire."
        ],
        [
            "Je suis une vraie voix."
        ],
        [
            "D\u00e9sol\u00e9 je ne pourrai pas venir."
        ],
        [
            "Mmm que c'est bon ce repas."
        ],
        [
            "Juste avant le vingt trois novembre deux mille vingt deux."
        ]
    ],
    "eval_split_max_size": null,
    "eval_split_size": 0.01,
    "use_speaker_weighted_sampler": false,
    "speaker_weighted_sampler_alpha": 1.0,
    "use_language_weighted_sampler": false,
    "language_weighted_sampler_alpha": 1.0,
    "use_length_weighted_sampler": false,
    "length_weighted_sampler_alpha": 1.0,
    "model_args": {
        "num_chars": 122,
        "out_channels": 513,
        "spec_segment_size": 32,
        "hidden_channels": 192,
        "hidden_channels_ffn_text_encoder": 768,
        "num_heads_text_encoder": 2,
        "num_layers_text_encoder": 6,
        "kernel_size_text_encoder": 3,
        "dropout_p_text_encoder": 0.1,
        "dropout_p_duration_predictor": 0.5,
        "kernel_size_posterior_encoder": 5,
        "dilation_rate_posterior_encoder": 1,
        "num_layers_posterior_encoder": 16,
        "kernel_size_flow": 5,
        "dilation_rate_flow": 1,
        "num_layers_flow": 4,
        "resblock_type_decoder": "1",
        "resblock_kernel_sizes_decoder": [
            3,
            7,
            11
        ],
        "resblock_dilation_sizes_decoder": [
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ],
            [
                1,
                3,
                5
            ]
        ],
        "upsample_rates_decoder": [
            8,
            8,
            2,
            2
        ],
        "upsample_initial_channel_decoder": 512,
        "upsample_kernel_sizes_decoder": [
            16,
            16,
            4,
            4
        ],
        "periods_multi_period_discriminator": [
            2,
            3,
            5,
            7,
            11
        ],
        "use_sdp": true,
        "noise_scale": 1.0,
        "inference_noise_scale": 0.667,
        "length_scale": 1.0,
        "noise_scale_dp": 1.0,
        "inference_noise_scale_dp": 0.8,
        "max_inference_len": null,
        "init_discriminator": true,
        "use_spectral_norm_disriminator": false,
        "use_speaker_embedding": false,
        "num_speakers": 0,
        "speakers_file": null,
        "d_vector_file": null,
        "speaker_embedding_channels": 256,
        "use_d_vector_file": false,
        "d_vector_dim": 0,
        "detach_dp_input": true,
        "use_language_embedding": false,
        "embedded_language_dim": 4,
        "num_languages": 0,
        "language_ids_file": null,
        "use_speaker_encoder_as_loss": false,
        "speaker_encoder_config_path": "",
        "speaker_encoder_model_path": "",
        "condition_dp_on_speaker": true,
        "freeze_encoder": false,
        "freeze_DP": false,
        "freeze_PE": false,
        "freeze_flow_decoder": false,
        "freeze_waveform_decoder": false,
        "encoder_sample_rate": null,
        "interpolate_z": true,
        "reinit_DP": false,
        "reinit_text_encoder": false
    },
    "lr_gen": 0.0002,
    "lr_disc": 0.0002,
    "lr_scheduler_gen": "ExponentialLR",
    "lr_scheduler_gen_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "lr_scheduler_disc": "ExponentialLR",
    "lr_scheduler_disc_params": {
        "gamma": 0.999875,
        "last_epoch": -1
    },
    "kl_loss_alpha": 1.0,
    "disc_loss_alpha": 1.0,
    "gen_loss_alpha": 1.0,
    "feat_loss_alpha": 1.0,
    "mel_loss_alpha": 45.0,
    "dur_loss_alpha": 1.0,
    "speaker_encoder_loss_alpha": 1.0,
    "return_wav": true,
    "use_weighted_sampler": false,
    "weighted_sampler_attrs": {},
    "weighted_sampler_multipliers": {},
    "r": 1,
    "num_speakers": 0,
    "use_speaker_embedding": false,
    "speakers_file": null,
    "speaker_embedding_channels": 256,
    "language_ids_file": null,
    "use_language_embedding": false,
    "use_d_vector_file": false,
    "d_vector_file": null,
    "d_vector_dim": 0,
    "restore_path": "/home/caraduf/VoiceCloning/Models/VITS_A_RTX-September-25-2022_02+49PM-0000000/best_model_56785.pth",
    "github_branch": "inside_docker"
}
```

**Reply by WeberJulian** [reply]
You should use phoneme_cleaners since it has english specific normalizations.

**Reply by WeberJulian** [reply]
And yeah I guess your best chance would be to add more data and enable speaker balancing to avoid overfitting with the small dataset.

**Reply by Ca-ressemble-a-du-fake** [reply]
Should or should NOT use phoneme cleaners for French ? if not which one do you advise ?

**Reply by WeberJulian** [reply]
You should use espeak phonemizer with multilingual_cleaner

**Reply by Ca-ressemble-a-du-fake** [reply]
Thanks for the advice @WeberJulian. I switched to multilingual_cleaners. By the way why do you replace ";", ";" and "-" by "," in the cleaners ?

Phonemizer is already set to "espeak" (see config.json). I noticed in another trial with a recipe that it was missing. Does it default to "espeak" if omitted ?

**Reply by WeberJulian** [reply]
Just easier to make it easier to learn.
Yes it defaults to espeak.

**Reply by Ca-ressemble-a-du-fake** [reply]
Great thank you !

**Reply by Ca-ressemble-a-du-fake** [reply]
Quick feedback from my side : using multilingual cleaner and unfrozen DP and encoder did the trick. I completely commented the `Characters=CharacterConfig` part in both original and fine tuned model. Thanks for your suggestion @WeberJulian !

**Reply by iamkhalidbashir** [reply]
So no freezing was done, and you produced intelligent speech from 562 samples from scratch? or were you fine-tuned?
And for fine-tuning do you include the previous dataset of the original model or just the dataset of new samples
Help is much appreciated! I have settled with yourtts as IMS Toucan and YourTTs are not a natural sound. Vits give natural sound but the voice is not of me when  fine tune with freezing (decoder and PE) and when I not freeze anything then I have a robotic metallic sound

**Reply by Ca-ressemble-a-du-fake** [reply]
I trained from scratch because I could not figure out how to correcly fine tune ! From my understanding if you only include your dataset you will lose the previous voices/accents/language.

**Reply by iamkhalidbashir** [reply]
Have you tried finetuning without freezing anything and using the previous dataset and the new dataset?
Also when training from scratch do you notice any metallic voices at the start of training? if yes then when does those noise goes off?

**Reply by Ca-ressemble-a-du-fake** [reply]
No I did not (I was running low on disk space so could not download all that datasets).
I think the metallic voice goes away around 20k-30k.
